### ISSUE TITLE
Store macros

### ISSUE THREAD

**Issue by ViktorWalter** [qutebrowser/qutebrowser#4385]
Is it possible to store macros somehow?
I am specifically exploiting macros to bind a command sequence starting with jseval (;; does not split next command, so I made a macro where I call one command with jseval and another command after that).
The problem is, that macros apparently don't carry over between sessions, so the feature does not stick.

**Comment by The-Compiler** [comment]
It's currently not possible to store macros. It's been requested in IRC before, but apparently there wasn't an issue about it yet.

**Comment by ViktorWalter** [comment]
Ok, thank you. Is there another way I can bind multiple commands starting with a "jseval [script]" statement?

**Comment by The-Compiler** [comment]
I don't think there is currently - I agree the current situation isn't optimal, see #1850/#2017.

**Comment by ViktorWalter** [comment]
Ok, just for the hell of it I tried to do this by inserting a \n into the binding in the config.py. I can report that reloading such config reliably crashes qutebrowser :-).

**Comment by The-Compiler** [comment]
I'll need more information than that - is it the same stacktrace as in #3580?

**Comment by ViktorWalter** [comment]
No, it differs. It catches the same warning as when I try to insert a hard nextline into the qutebrowser command line "may not contain unprintable chars!" , but apparently when forced this way it can't deal with it. Though I understand that this is just an edge case of the user going way off the rails:
Traceback (most recent call last):
  File "/home/viktor/git/qutebrowser/qutebrowser/commands/runners.py", line 325, in run_safely
    self.run(text, count)
  File "/home/viktor/git/qutebrowser/qutebrowser/commands/runners.py", line 304, in run
    result.cmd.run(self._win_id, args, count=count)
  File "/home/viktor/git/qutebrowser/qutebrowser/commands/command.py", line 507, in run
    self.handler(*posargs, **kwargs)
  File "/home/viktor/git/qutebrowser/qutebrowser/config/configcommands.py", line 384, in config_source
    configfiles.read_config_py(filename)
  File "/home/viktor/git/qutebrowser/qutebrowser/config/configfiles.py", line 586, in read_config_py
    api.finalize()
  File "/home/viktor/git/qutebrowser/qutebrowser/config/configfiles.py", line 377, in finalize
    self._config.update_mutables()
  File "/home/viktor/git/qutebrowser/qutebrowser/config/config.py", line 455, in update_mutables
    self.set_obj(name, new_value, save_yaml=save_yaml)
  File "/home/viktor/git/qutebrowser/qutebrowser/config/config.py", line 399, in set_obj
    self._set_value(opt, value, pattern=pattern)
  File "/home/viktor/git/qutebrowser/qutebrowser/config/config.py", line 283, in _set_value
    opt.typ.to_py(value)  # for validation
  File "/home/viktor/git/qutebrowser/qutebrowser/config/configtypes.py", line 1291, in to_py
    for key, val in value.items()}
  File "/home/viktor/git/qutebrowser/qutebrowser/config/configtypes.py", line 1291, in <dictcomp>
    for key, val in value.items()}
  File "/home/viktor/git/qutebrowser/qutebrowser/config/configtypes.py", line 1291, in to_py
    for key, val in value.items()}
  File "/home/viktor/git/qutebrowser/qutebrowser/config/configtypes.py", line 1291, in <dictcomp>
    for key, val in value.items()}
  File "/home/viktor/git/qutebrowser/qutebrowser/config/configtypes.py", line 888, in to_py
    self._basic_py_validation(value, str)
  File "/home/viktor/git/qutebrowser/qutebrowser/config/configtypes.py", line 173, in _basic_py_validation
    self._basic_str_validation(value)
  File "/home/viktor/git/qutebrowser/qutebrowser/config/configtypes.py", line 189, in _basic_str_validation
    value, "may not contain unprintable chars!")
qutebrowser.config.configexc.ValidationError: Invalid value 'jseval --quiet function elementInViewport2(el) { var top = el.offsetTop; var left = el.offsetLeft; var width = el.offsetWidth; var height = el.offsetHeight; while(el.offsetParent) { el = el.offsetParent; top += el.offsetTop; left += el.offsetLeft; } return ( top < (window.pageYOffset + window.innerHeight) && left < (window.pageXOffset + window.innerWidth) && (top + height) > window.pageYOffset && (left + width) > window.pageXOffset); } var pres = document.getElementsByTagName('pre'); var focusedAny=false; for(var i=0; i<pres.length; i++) { if( pres[i].focused ){ focusedAny =true; var focusedPrev =i; } } var focusedNext = 0; if (focusedAny){ focusedNext = focusedPrev+1; if (focusedNext >= (pres.length-1)){ focusedNext=0; } } else{ var focusedPrev = pres.length-1; focusedNext = 0; } var foundAnyVisible=false; for(var j=0; j<pres.length; j++){ if (focusedNext==focusedPrev){ break; } if (elementInViewport2(pres[focusedNext])){ foundAnyVisible =true; break; } focusedNext=focusedNext+1; if (focusedNext >= (pres.length-1)){ focusedNext=0; } } if (foundAnyVisible){ pres[focusedPrev].focused=false; pres[focusedPrev].style="border:1px solid White;"; pres[focusedPrev].id=""; pres[focusedNext].focused=true; pres[focusedNext].id="clickhere"; pres[focusedNext].style="border:2px solid Tomato;"; } 
;; click-element id clickhere' - may not contain unprintable chars!

**Comment by The-Compiler** [comment]
I do think that's the same underlying issue. Going to hide the comments here as off-topic now.

**Comment by ViktorWalter** [comment]
Ok. Sorry for the spam.

**Comment by ViktorWalter** [comment]
For the interested, I found a hacky workaround that I am using now:
I bind `fake-key --global` with a set of other keybindings forming a pseudo-macro. These other keybindings are something weird, such as `_%_` or `_$_` that is unlikely to be pressed separately anyway, and contain commands that can't be chained otherwise, so the whole binding looks for instance like `fake-key --global _%__$_`.
This can be stored in `config.py` normally.
